server/keycloak.ts

```diff
@@ -373,6 +402,65 @@ class Keycloak {
       throw error;
     }
   }
+
+  /**
+   * Migrate user's old keycloak id to new keycloak id and update roles
+   *
+   * @param keycloakId
+   * @param email
+   * @param username
+   */
+  async migrateUser(keycloakId: string, email: string, username: string): Promise<string[] | null> {
+    const migrationStatus = await dbClient.db[collections.USER_MIGRATION].findOne({
+      'email ilike': email,
+      'username ilike': username.includes('@bceid') ? `${username.split('@')[0]}@bceid%` : username,
+    });
+
+    if (!migrationStatus || migrationStatus.status !== 'pending') {
+      return null;
+    }
+
+    const undefinedRoles = migrationStatus.roles.filter((role) => !this.roleIdMap[role]);
+
+    if (undefinedRoles.length > 0) {
+      const msg = `${undefinedRoles.join(', ')} not defined`;
+      logger.error(msg, { context: 'kc-user-migration' });
+      throw Error(msg);
+    }
+
+    try {
+      await this.setUserRoles(keycloakId, migrationStatus.roles);
+    } catch (e) {
+      logger.error(`failed to update user(${keycloakId})'s role`, { context: 'kc-user-migration' });
+      throw e;
+    }
+
+    const [user] = await dbClient.db[collections.USERS].findDoc({
+      'userInfo.username ilike': username.includes('@bceid')
+        ? `${username.split('@')[0]}@bceid%`
+        : username,
+      'userInfo.email ilike': email,
+    });
+    if (user) {
+      user.userInfo.username = username;
+      await dbClient.db[collections.USERS].updateDoc(user.id, {
```

Comment:
Will we want to reuse some of the work from your migration script to update id and participant status employer_ids?

Reply:
I hope HCAP-1484 to be merged before this so that we don't need to update the `participant_status` table.